### DISCUSSION TITLE
Async does not perform the tasks asynronously.

### DISCUSSION THREAD

**Discussion by marek22k** [socketry/async#234] [Q&A]
Hello,

I have a code which should read for an IPv4 and IPv6 from two files and then write the results to a SQLite database. This offers to do synron. With thread it works, but not with async.
```
require 'timeout'
require 'async'
require 'async/barrier'
require 'sqlite3'
require_relative 'parser'
require_relative 'relation_builder'

#Timeout::timeout(2000) do
    Async do |task|
        puts "Preparing to parse."

        # Create objects for IPv4 mrt file
        file4 = File.new 'master4_latest.mrt', 'rb'
        rbuilder4 = RelationBuilder.new
        parser4 = MrtParser.new file4, rbuilder4

        # Create objects for IPv6 mrt file
        file6 = File.new 'master6_latest.mrt', 'rb'
        rbuilder6 = RelationBuilder.new
        parser6 = MrtParser.new file6, rbuilder6

        barrier = Async::Barrier.new

        barrier.async do
            puts "Parse IPv6."
            STDOUT.flush

            # Parse IPv6 mrt file
            parser6.parse
            rbuilder6.build_as_origins
        end
    
        barrier.async do
            puts "Parse IPv4."
            STDOUT.flush
        
            # Parse IPv4 mrt file
            parser4.parse
            rbuilder4.build_as_origins
        end.wait

        barrier.wait

        puts "Connect to SQLite database."
        db = SQLite3::Database.new 'data.db'

        barrier.async do
            puts "Write contibutors to database."
            db.execute 'DROP TABLE IF EXISTS "contributors"'
            db.execute 'CREATE TABLE "contributors" ( "asn" INTEGER NOT NULL UNIQUE, PRIMARY KEY("asn") );'
            contributors = (parser4.meta[:contributors] + parser6.meta[:contributors]).uniq
            contributors.each do |asn|
                db.execute "INSERT INTO contributors (asn) VALUES (#{asn});"
            end
        end

        barrier.async do
            puts "Write IPv4 Prefix<->ASN relations to database."
            db.execute 'DROP TABLE IF EXISTS "prefix_origins4"'
            db.execute 'CREATE TABLE "prefix_origins4" ( "prefix" TEXT NOT NULL UNIQUE, "origin" TEXT NOT NULL, PRIMARY KEY("prefix") );'
            rbuilder4.prefix_origins.each do |prefix, ass|
                db.execute "INSERT INTO prefix_origins4 (prefix, origin) VALUES (\"#{prefix}/#{prefix.prefix}\", \"#{ass.join(',')}\");"
            end
        end

        barrier.async do
            puts "Write IPv6 Prefix<->ASN relations to database."
            db.execute 'DROP TABLE IF EXISTS "prefix_origins6"'
            db.execute 'CREATE TABLE "prefix_origins6" ( "prefix" TEXT NOT NULL UNIQUE, "origin" TEXT NOT NULL, PRIMARY KEY("prefix") );'
            rbuilder6.prefix_origins.each do |prefix, ass|
                db.execute "INSERT INTO prefix_origins6 (prefix, origin) VALUES (\"#{prefix}/#{prefix.prefix}\", \"#{ass.join(',')}\");"
            end
        end

        barrier.async do
            puts "Write ASN<->ASN for IPv4 relations to database."
            db.execute 'DROP TABLE IF EXISTS "relations4"'
            db.execute 'CREATE TABLE "relations4" ( "asn" INTEGER NOT NULL UNIQUE, "neighbors" TEXT NOT NULL, PRIMARY KEY("asn") );'
            rbuilder4.relations.each do |asn, ass|
                db.execute "INSERT INTO relations4 (asn, neighbors) VALUES (#{asn}, \"#{ass.join(',')}\");"
            end
        end

        barrier.async do
            puts "Write ASN<->ASN for IPv6 relations to database."
            db.execute 'DROP TABLE IF EXISTS "relations6"'
            db.execute 'CREATE TABLE "relations6" ( "asn" INTEGER NOT NULL UNIQUE, "neighbors" TEXT NOT NULL, PRIMARY KEY("asn") );'
            rbuilder6.relations.each do |asn, ass|
                db.execute "INSERT INTO relations6 (asn, neighbors) VALUES (#{asn}, \"#{ass.join(',')}\");"
            end
        end

        #File.write('prefix_origins.json', JSON.dump(rbuilder4.prefix_origins.transform_keys{|key| "#{key}/#{key.prefix}"}))
        #File.write('as_origins.json', JSON.dump(rbuilder4.as_origins.transform_values{|value| value.map{|key|"#{key}/#{key.prefix}"}}))
        #File.write('relations.json', JSON.dump(rbuilder4.relations))
        #File.write('meta.json', JSON.dump(meta.transform_keys(&:to_s)))

        barrier.wait

        puts "Close database."
        db.close
    end
#end
```
Actually, "Parse IPv6." and "Parse IPv4." should be output at the same time at the beginning, but this does not happen. If I use the native Ruby threads this happens. What am I doing wrong here?

**Comment by ioquatix** [comment]
A few thoughts:

- SQLite3 gem may do a poor job of releasing the GVL in some cases.
- File IO is only asynchronous when using `io_uring` backend, in other cases it's synchronous.
- SQLite3 doesn't use the fiber scheduler hooks at this time - it may be possible to fix that.

Do you have any workload in this script that is predominantly IO bound? i.e. what is the performance issue you are hoping to address?

**Comment by marek22k** [comment]
Thanks for your answer.

I plan to switch from SQLite3 to ActiveSupport soon.
However, the parsing itself has nothing to do with the database. The parsing reads a file (or one for IPv4 and another for IPv6) and processes it with the "unpack" function of Ruby. My hope was that by using async, the parsing of the IPv4 file and the IPv6 file could happen at the same time, so it would only take a little half the time.

**Reply by ioquatix** [reply]
How long does parsing take?

Parsing sounds mostly CPU bound not IO bound.

In the case it's CPU bound, the best way to is to use multiple processes. You can do this easily using `async-container` if that's something you are interested in: https://github.com/socketry/async-container/blob/main/examples/container.rb

It won't matter to use AR or SQLite3, unless by using AR you are using Postgres and are IO bound. But I imagine your workload is mostly CPU bound.

**Reply by marek22k** [reply]
> How long does parsing take?

```
real	0m49,943s
user	0m49,219s
sys	0m0,337s
```
> 
> Parsing sounds mostly CPU bound not IO bound.
> 
> In the case it's CPU bound, the best way to is to use multiple processes. You can do this easily using `async-container` if that's something you are interested in: https://github.com/socketry/async-container/blob/main/examples/container.rb

This looks good. Thanks for the tip! I'm still quite new to asynchronous programs: what's the difference between an additional thread and an additional process?
> 
> It won't matter to use AR or SQLite3, unless by using AR you are using Postgres and are IO bound. But I imagine your workload is mostly CPU bound.

I also read all the data from the files first and then save it to the database.

**Reply by marek22k** [reply]
```
I, [2023-06-02T11:00:14.090864 #22742]  INFO -- : Preparing to parse.
I, [2023-06-02T11:00:14.095405 #22744]  INFO -- : Parse IPv6.
I, [2023-06-02T11:00:14.096907 #22745]  INFO -- : Parse IPv4.
  0.0s     warn: Async::Task [oid=0x424] [ec=0x438] [pid=22742] [2023-06-02 11:08:14 +0200]
               | Task may have ended with unhandled exception.
               |   Timeout::Error: execution expired
               |   → /home/marek/.rvm/gems/ruby-3.2.2/gems/async-container-0.16.12/lib/async/container/group.rb:155 in `select'
               |     /home/marek/.rvm/gems/ruby-3.2.2/gems/async-container-0.16.12/lib/async/container/group.rb:155 in `wait_for_children'
               |     /home/marek/.rvm/gems/ruby-3.2.2/gems/async-container-0.16.12/lib/async/container/group.rb:74 in `wait'
               |     /home/marek/.rvm/gems/ruby-3.2.2/gems/async-container-0.16.12/lib/async/container/generic.rb:106 in `wait'
               |     backend/relations/main.rb:183 in `block (2 levels) in <main>'
               |     /home/marek/.rvm/gems/ruby-3.2.2/gems/async-2.5.1/lib/async/task.rb:158 in `block in run'
               |     /home/marek/.rvm/gems/ruby-3.2.2/gems/async-2.5.1/lib/async/task.rb:310 in `block in schedule'

real	8m0,620s
user	0m0,419s
sys	0m0,085s
```
With container (judging by the output) both happens at the same time, but exceeds the time frame.

**Reply by ioquatix** [reply]
Do you mind sharing your updated code? (or putting it in a repo)

**Reply by marek22k** [reply]
It is one part of a larger program. [main.rb](https://codeberg.org/mark22k/BgpLookup/src/commit/afa5e657f8ea8f32e7b1b9446689e87184127feb/backend/relations/main.rb#L169)
Sample mrt file: 
[mrts.zip](https://github.com/socketry/async/files/11634088/mrts.zip)
Sample run command (in git root): `ruby backend/relations/main.rb -4 ~/Downloads/master4_latest.mrt.bz2 -6 ~/Downloads/master6_latest.mrt.bz2`

**Comment by ioquatix** [comment]
@ko1 do you have any thoughts about how to write such a program with Ruby? `Ractor` seems like the only solution here, unless we move to TruffleRuby with threads that have parallelism.
